File: GFG POTD/022-September-2025/030-generate-all-binary-strings.cpp
// https://www.geeksforgeeks.org/problems/generate-all-binary-strings/1/




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    // TC : O(N * 2^N)
    // SC : O(N) Recursive Depth
    void solve(int n, string curr, vector<string> &res) {
        if (curr.length() == n) {
            res.push_back(curr);
            return;
        }
        
        solve(n, curr + '0', res);
        solve(n, curr + '1', res);
    }
  public:
    // TC : O(N * 2^N)
    // SC : O(N)
    vector<string> binstr(int n) {
        vector<string> result;
        string curr;
        solve(n, curr, result);
        return result;
    }
};

File: GFG POTD/022-September-2025/030-generate-all-binary-strings.java
// https://www.geeksforgeeks.org/problems/generate-all-binary-strings/1/




import java.util.*;

class Solution {
    public ArrayList<String> binstr(int n) {
        ArrayList<String> result = new ArrayList<>();
        StringBuilder str = new StringBuilder();
        solve(n, str, result);
        return result;
    }
    
    private static void solve(int n, StringBuilder str, ArrayList<String> result) {
        if (n == 0) {
            result.add(str.toString());
            return;
        }
        
        str.append('0');
        solve(n - 1, str, result);
        str.deleteCharAt(str.length() - 1);
        
        str.append('1');
        solve(n - 1, str, result);
        str.deleteCharAt(str.length() - 1);
    }
}